### PR TITLE
Add interactive map to home page

### PR DESCRIPTION
**Pull request by AaronLopes**
(project ID update in this PR as well)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 target/
 .idea
+portfolio/src/main/webapp/config.js
```
Comment (anbenson) [inline .gitignore:3]: nit: Ending newline.

File: portfolio/src/main/webapp/index.html [hunk 2]
```diff
@@ -6,7 +6,7 @@
     <link href="style.css" rel="stylesheet">
     <link href="https://fonts.googleapis.com/css2?family=Roboto:wght@300;400;700&display=swap" rel="stylesheet">
   </head>
-  <body>
+  <body>
```
Comment (anbenson) [inline portfolio/src/main/webapp/index.html:9]: nit: remove trailing whitespace

File: portfolio/src/main/webapp/index.html [hunk 3]
```diff
@@ -27,7 +31,7 @@ <h2><div id="changeText"></div></h2>
       <input type="number" id="comment-choice" min="1" max="10" value="1">
       <button id="commentBtn">Display</button>
       <ul id="comment-container"></ul>
-    </div>
-    <script src="script.js"></script>
+    </div>    
   </body>
+  <script src="script.js"></script>
```
Comment (anbenson) [inline portfolio/src/main/webapp/index.html:36]: It felt very weird to me to put this outside of body: https://stackoverflow.com/questions/3037725/is-it-wrong-to-place-the-script-tag-after-the-body-tag

IMO, best practice is to put it in head but make it defer or async, but no need to do that here.

File: portfolio/src/main/webapp/index.html [hunk 4]
```diff
@@ -27,7 +31,7 @@ <h2><div id="changeText"></div></h2>
       <input type="number" id="comment-choice" min="1" max="10" value="1">
       <button id="commentBtn">Display</button>
       <ul id="comment-container"></ul>
-    </div>
-    <script src="script.js"></script>
+    </div>
```
Comment (anbenson) [inline portfolio/src/main/webapp/index.html:34]: I'm so lost what got changed on this line lol

File: portfolio/src/main/webapp/script.js [hunk 5]
```diff
@@ -11,7 +11,6 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
-
```
Comment (anbenson) [inline portfolio/src/main/webapp/script.js:14]: Did you mean to delete this? Seems you'd want separation between the license header and your method documentation.

File: portfolio/src/main/webapp/script.js [hunk 6]
```diff
@@ -22,6 +21,47 @@ var counter = 1;
 var intervalLength = setInterval(change, 2000);
 document.getElementById('commentBtn').addEventListener('click', getComments);
 
+var script = document.createElement('script');
```
Comment (anbenson) [inline portfolio/src/main/webapp/script.js:24]: Instead of putting this in via JS, can you just put it in statically in the HTML?

File: portfolio/src/main/webapp/script.js [hunk 7]
```diff
@@ -22,6 +21,47 @@ var counter = 1;
...
+script.src = `https://maps.googleapis.com/maps/api/js?key=${MAPKEY}&callback=initMap`;
```
Comment (anbenson) [inline portfolio/src/main/webapp/script.js:25]: I'm assuming MAPKEY is meant to be replaced by the developer using this?

File: portfolio/src/main/webapp/script.js (hunk 6 above)
Comment (anbenson) [reply portfolio/src/main/webapp/script.js:24]: If it doesn't work, you might be hitting a subtle race condition that the tutorials never discuss.

Background: HTML parsers (e.g. browser) process HTML in source code order. When they hit a script tag, they stop parsing until the script downloads and executes, by default. (If defer or async is present, they may continue accordingly, see the docs.)

Race condition: This gmaps script tag in the head is downloaded but since it's async, the browser keeps going until it hits the end of body and executes this script we're in which defines initMap. The race condition is which script executes first - this one needs to execute first to define initMap for the gmaps script not to error. The 2nd time onward, it'll probably always error since the gmaps script will be cached and will execute near immediately. So consider ordering your script tags accordingly to avoid this.

File: portfolio/src/main/webapp/script.js [hunk 8]
```diff
@@ -22,6 +21,47 @@ var counter = 1;
...
+script.defer = true;
+script.async = true;
+
+window.initMap = function() {
+    const map = new google.maps.Map(
+      document.getElementById('map'),
+      {center: {lat: 37.422403, lng: -122.088073}, zoom: 5});
+    addLandmark(
+      map, 37.3230, -122.0322, 'Home',
+      'Where I currently live, and spent my high school years. Home to Apple HQ, the best boba, and my closest friends :)');
```
Comment (anbenson) [inline portfolio/src/main/webapp/script.js:35]: what's the best boba

i'm ready

File: portfolio/src/main/webapp/script.js [hunk 9]
```diff
@@ -22,6 +21,47 @@ var counter = 1;
...
 document.getElementById('commentBtn').addEventListener('click', getComments);
 
+var script = document.createElement('script');
+script.src = `https://maps.googleapis.com/maps/api/js?key=${MAPKEY}&callback=initMap`;
+script.defer = true;
+script.async = true;
+
+window.initMap = function() {
+    const map = new google.maps.Map(
+      document.getElementById('map'),
```
Comment (anbenson) [inline portfolio/src/main/webapp/script.js:31]: The indentation here is pretty hard to read.

File: portfolio/src/main/webapp/script.js [hunk 10]
```diff
@@ -22,6 +21,47 @@ var counter = 1;
...
+      {center: {lat: 37.422403, lng: -122.088073}, zoom: 5});
+    addLandmark(
```
Comment (anbenson) [inline portfolio/src/main/webapp/script.js:33]: I would put all these datapoints in a list of named variables somehow and call addLandmark in a loop. Right now the data is embedded in the code, and generally you want to separate data and code.

File: portfolio/src/main/webapp/script.js [hunk 11]
```diff
@@ -22,6 +21,47 @@ var counter = 1;
...
+document.head.appendChild(script);
+
+function addLandmark(map, lat, lng, title, description) {
+  const marker = new google.maps.Marker(
+      {position: {lat: lat, lng: lng}, map: map, title: title});
+
+  const infoWindow = new google.maps.InfoWindow({content: description});
+  marker.addListener('click', () => {
+    infoWindow.open(map, marker);
+    var latLng = new google.maps.LatLng(lat, lng);
```
Comment (anbenson) [inline portfolio/src/main/webapp/script.js:59]: Avoid var.

File: walkthroughs/week-4-libraries/maps/examples/google-tour/src/main/webapp/index.html [hunk 12]
```diff
@@ -2,7 +2,7 @@
 <html>
 <head>
   <title>Google Tour</title>
-  <script src="https://maps.googleapis.com/maps/api/js?key=YOUR_API_KEY"></script>
```
Comment (anbenson) [inline walkthroughs/week-4-libraries/maps/examples/google-tour/src/main/webapp/index.html:5]: Revert changes to walkthroughs directory?

File: portfolio/src/main/webapp/script.js [hunk 13]
```diff
@@ -22,6 +21,47 @@ var counter = 1;
 var intervalLength = setInterval(change, 2000);
```
Comment (joanchenx) [inline portfolio/src/main/webapp/script.js:21]: I didn't see this earlier but for all of your `var` usages, could you switch to `const` or `let` instead?

File: portfolio/src/main/webapp/script.js (hunk 8 above)
Comment (AaronLopes) [reply portfolio/src/main/webapp/script.js:35]: ten ren is the best I've had 🤷‍♂️

File: portfolio/src/main/webapp/script.js (hunk 7 above)
Comment (AaronLopes) [reply portfolio/src/main/webapp/script.js:25]: yup!

File: portfolio/src/main/webapp/script.js [hunk 14]
```diff
@@ -16,12 +16,41 @@
...
+    for (let i = 0; i < titles.length; i++) {
+        addLandmark(map, lats[i], lngs[i], titles[i], descriptions[i])
+    }
+};
+
+function addLandmark(map, lat, lng, title, description) {
+  const marker = new google.maps.Marker(
+      {position: {lat: lat, lng: lng}, map: map, title: title});
+
+  const infoWindow = new google.maps.InfoWindow({content: description});
```
Comment (joanchenx) [inline portfolio/src/main/webapp/script.js:45]: Up to you, but do you want to add the title to infoWindow?

File: portfolio/src/main/webapp/script.js [hunk 15]
```diff
@@ -16,12 +16,41 @@
...
+    const titles = ['Home', 'UMass Amherst', 'Georgia Tech', 'Home Away From Home', 'An Extraordinary Layover'];
+    const lats = [37.3230, 42.3868, 33.7756, 19.4564, 51.5074];
+    const lngs = [-122.0322,-72.5301, -84.3963, 72.7925, -0.1278]
+    const descriptions = ['Where I currently live, and spent my high school years. Home to Apple HQ, the best boba, \
+    and my closest friends :)', 'Snowy, remote, but cozy; where I spent my first 1.5 years of college.', 'Where I \
+    currently attend college!', 'Where my parents grew up; the comfort of my extended family and the cities unique \
+    culture always brings me back', 'Spent an ambitious 8 hours searching for all the famous landmarks, check out pictures in misc!'];
+    for (let i = 0; i < titles.length; i++) {
+        addLandmark(map, lats[i], lngs[i], titles[i], descriptions[i])
+    }
```
Comment (joanchenx) [inline portfolio/src/main/webapp/script.js:38]: I think have these all be separate arrays makes it a bit hard to parse which titles match with which lats/lngs/descriptions.

What about add a new class, something like:
```
class Landmark {
  constructor(title, lat, lng, description) {
      this.title = title;
      this.lat = lat;
      this.lng = lng;
      this.description = description;
  }
}
```

And then you can create new Landmarks with: `new Landmark('Home', 37.3230, -122.0322, 'Where I..')` and put all your Landmarks into a single array that you iterate through. Then, you can also change `addLandmark` to take in just two params: `function addLandmark(map, landmark)`

File: portfolio/src/main/webapp/script.js [hunk 16]
```diff
@@ -16,12 +16,51 @@
  * Cycles through greetings in different languages for front page. 
  */
 const greetings =['Hi, nice to meet you!', '¡Hola, encantado de conocerte!', '你好，很高興見到你', 'Salut! Enchanté!'];
-var textElement = document.getElementById('changeText');
+let textElement = document.getElementById('changeText');
```
Comment (anbenson) [inline portfolio/src/main/webapp/script.js:19]: const

Comment (anbenson) [reply portfolio/src/main/webapp/script.js:19]: I'm going to guess the choice of `let` was motivated by the fact that textElement is mutated on the next line. The distinction is subtle - const does not refer to the object in question being constant, but the _binding_ being constant e.g. the variable `textElement` always refers to this object and will never refer to anything else. The underlying object can still be mutated e.g. a const array can have elements added and removed from it.

File: portfolio/src/main/webapp/script.js [hunk 17]
```diff
@@ -16,12 +16,51 @@
...
+
+  marker.addListener('click', () => {
+    infoWindow.open(map, marker);
+    let latLng = new google.maps.LatLng(Landmark.lat, Landmark.lng);
+    map.panTo(latLng);
+    map.setZoom(10);
+  });
+}
+
 function change() {
```
Comment (anbenson) [inline portfolio/src/main/webapp/script.js:64]: This function is sitting in the middle of your other functions, but it's not related to them. Can you move it out, maybe near the code that calls it?

File: portfolio/src/main/webapp/script.js [hunk 18]
```diff
@@ -16,12 +16,51 @@
...
+    this.description = description;     
+  }
+}
+
+function initMap() {
+    const map = new google.maps.Map(
+      document.getElementById('map'),
+      {center: {lat: 37.422403, lng: -122.088073}, zoom: 5});
+    
+    let landmarks = [new Landmark('Home', 37.3230, -122.0322, 'Where I currently live, and spent my high school years. \
```
Comment (anbenson) [inline portfolio/src/main/webapp/script.js:39]: const

File: portfolio/src/main/webapp/script.js [hunk 19]
```diff
@@ -16,12 +16,51 @@
...
+    this.lat = lat;
+    this.lng = lng;
+    this.description = description;     
+  }
+}
+
+function initMap() {
+    const map = new google.maps.Map(
+      document.getElementById('map'),
+      {center: {lat: 37.422403, lng: -122.088073}, zoom: 5});
```
Comment (anbenson) [inline portfolio/src/main/webapp/script.js:37]: May I suggest

```js
const map = new google.maps.Map(
  document.getElementById('map'),
  {
    center: {
      lat: 37.422403,
      lng: -122.088073,
    },
    zoom: 5,
  }
);
```
(note trailing commas)

or at the very least
```js
const map = new google.maps.Map(
  document.getElementById('map'),
  {center: {lat: 37.422403, lng: -122.088073}, zoom: 5}
);
```

File: portfolio/src/main/webapp/script.js (hunk 18 above)
Comment (anbenson) [inline portfolio/src/main/webapp/script.js:39]: And again, formatting:

```js
const landmarks = [
  new Landmark(...),
  new Landmark(...),
  ...
  new Landmark(...),
];
```

(note ending semicolon, which was missing)

File: portfolio/src/main/webapp/script.js [hunk 20]
```diff
@@ -16,12 +16,51 @@
...
+    
+    let landmarks = [new Landmark('Home', 37.3230, -122.0322, 'Where I currently live, and spent my high school years. \
+    Home to Apple HQ, the best boba, and my closest friends :)'), new Landmark('UMass Amherst', 42.3868, -72.5301, 'Snowy, \
+    remote, but cozy; where I spent my first 1.5 years of college.'), new Landmark('Georgia Tech', 33.7756, -84.3963, 'Where I \
+    currently attend college!'), new Landmark('Home, Away From Home', 19.4564, 72.7925, 'Where my parents grew up; the comfort \
+    of my extended family and the cities unique culture always brings me back'), new Landmark('An Extraordinary Layover',
+    51.5074, -0.1278, 'Spent an ambitious 8 hours searching for all the famous landmarks, check out pictures in misc!') ]
+
+    for (let landmark of landmarks) {
+        addLandmark(map, landmark)
```
Comment (anbenson) [inline portfolio/src/main/webapp/script.js:47]: missing semicolon

File: portfolio/src/main/webapp/script.js [hunk 21]
```diff
@@ -16,12 +16,51 @@
...
+    }
+};
+
+function addLandmark(map, Landmark) {
```
Comment (anbenson) [inline portfolio/src/main/webapp/script.js:51]: variable names should not be capitalized (Landmark)

File: portfolio/src/main/webapp/script.js [hunk 22]
```diff
@@ -16,12 +16,51 @@
...
+  const marker = new google.maps.Marker(
+      {position: {lat: Landmark.lat, lng: Landmark.lng}, map: map, title: Landmark.title});
+  const infoWindow = new google.maps.InfoWindow({content: Landmark.description});
+
+  marker.addListener('click', () => {
```
Comment (anbenson) [inline portfolio/src/main/webapp/script.js:56]: Consider closing previous markers? If one is open when I click another one, they're both open which is a bit cluttered when I zoom out. But it's subjective - whatever you want the experience to be.
